README.md
> This file is part of Signifix, see <https://qu1x.org/signifix>.
> 
> Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
> 
> Signifix is free software: you can redistribute it and/or modify
> it under the terms of the GNU Lesser General Public License version 3
> as published by the Free Software Foundation on 29 June 2007.
> 
> Signifix is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
> GNU Lesser General Public License for more details.
> 
> You should have received a copy of the GNU Lesser General Public License
> along with Signifix. If not, see <https://www.gnu.org/licenses>.

# libsignifix

Number Formatter of Fixed Significance with Metric Unit Prefix

[![Build Status](https://travis-ci.org/qu1x/libsignifix.svg?branch=master)](https://travis-ci.org/qu1x/libsignifix)

[Documentation](https://qu1x.org/file/libsignifix-0.2.0/index.html)

Signifix guarantees a fixed number of significant figures and a fixed number
of resulting characters by automatically choosing the metric unit prefix and
appropriately adjusting the floating point precision.

This is a `C99` compatible library (`signifix.h`, `signifix.c`) with a `C++98`
compatible header only wrapper (`signifix.hxx`).

# Examples

Following examples make use of `C++14` features.

The fixed number of significant figures and resulting characters prevent the
digits and prefixed units from jumping to the left or right:

```c++
#include <signifix/signifix.hxx>

#include <cassert>

int main() {
	assert(signifix::format(1e-04) == "100.0 µ");
	assert(signifix::format(1e-03) == "1.000 m");
	assert(signifix::format(1e-02) == "10.00 m");
	assert(signifix::format(1e-01) == "100.0 m");
	assert(signifix::format(1e+00) == "1.000  ");
	assert(signifix::format(1e+01) == "10.00  ");
	assert(signifix::format(1e+02) == "100.0  ");
	assert(signifix::format(1e+03) == "1.000 k");
	assert(signifix::format(1e+04) == "10.00 k");
	assert(signifix::format(1e+05) == "100.0 k");
	assert(signifix::format(1e+06) == "1.000 M");
}
```

This is useful to smoothly refresh a transfer rate inside a terminal:

```c++
#include <signifix/signifix.hxx>

#include <cassert>

int main() {
	auto format_rate = [](auto bytes, auto seconds) {
		return signifix::format((double)bytes / (double)seconds) + "B/s";
	};

	assert(format_rate(42'667, 300) == "142.2  B/s");
	assert(format_rate(42'667,  30) == "1.422 kB/s");
	assert(format_rate(42'667,   3) == "14.22 kB/s");
}
```

Or to monitor a measured quantity like an electrical current including its
direction with an optional space placeholder to align with negative values.

```c++
#include <signifix/signifix.hxx>

#include <cassert>

int main() {
	auto format_load = [](auto current) {
		return signifix::format(current, signifix::SPACED) + "A";
	};

	assert(format_load( 1.476e-06) == " 1.476 µA");
	assert(format_load(-2.927e-06) == "-2.927 µA");
}
```

In case of displaying a file size difference after modification, a plus sign
might be preferred for positive values:

```c++
#include <signifix/signifix.hxx>

#include <cassert>

int main() {
	auto format_diff = [](auto curr, auto prev) {
		return signifix::format(curr - prev, signifix::SIGNED) + "B";
	};

	assert(format_diff(78'346, 57'393) == "+20.95 kB");
	assert(format_diff(27'473, 46'839) == "-19.37 kB");
}
```

# Installation

## Either getting a release

1. Download stable source distribution tarball.

		wget https://qu1x.org/file/libsignifix-0.2.0.tar.xz

2. Extract and enter.

		tar -xJf libsignifix-0.2.0.tar.xz
		cd libsignifix-0.2.0

## Or getting a snapshot

Required dependencies:

		git autoreconf libtool

1. Clone repository.

		git clone https://github.com/qu1x/libsignifix.git

2. Enter and generate latest source distribution.

		cd libsignifix
		./bootstrap

## Installing one of them

Required dependencies:

		build-essential

Optional dependencies:

		doxygen

1. Configure, build, and install.

To not generate documentation with `doxygen`, disable it with `--disable-doc`.

		./configure CFLAGS="$CFLAGS -std=c99"
		make
		sudo make install

2. Keep to uninstall someday.

		sudo make uninstall

# Version

libsignifix-0.2.0 <https://qu1x.org/signifix>

# License

GNU Lesser General Public License version 3

# Authors

* Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>



NEWS
2016-12-08

 * 0.2.0: Added tests and examples.
          Fixed missing rounding of significand.


2016-12-04

 * 0.1.0: Initial release

# This file is part of Signifix, see <https://qu1x.org/signifix>.
#
# Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
#
# Signifix is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License version 3
# as published by the Free Software Foundation on 29 June 2007.
#
# Signifix is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with Signifix. If not, see <https://www.gnu.org/licenses>.


tests/fixed_significance.cxx
// This file is part of Signifix, see <https://qu1x.org/signifix>.
//
// Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
//
// Signifix is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License version 3
// as published by the Free Software Foundation on 29 June 2007.
//
// Signifix is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with Signifix. If not, see <https://www.gnu.org/licenses>.

#include <signifix/signifix.hxx>

#include <cassert>

int main() {
	assert(signifix::format(1e+02) == "100.0  ");
	assert(signifix::format(1e+03) == "1.000 k");
	assert(signifix::format(1e+04) == "10.00 k");
	assert(signifix::format(1e+05) == "100.0 k");
	assert(signifix::format(1e+06) == "1.000 M");
}


tests/formatting_options.cxx
// This file is part of Signifix, see <https://qu1x.org/signifix>.
//
// Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
//
// Signifix is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License version 3
// as published by the Free Software Foundation on 29 June 2007.
//
// Signifix is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with Signifix. If not, see <https://www.gnu.org/licenses>.

#include <signifix/signifix.hxx>

#include <cassert>

int main() {
	assert(signifix::format(-1e+00, signifix::PROPER) == "-1.000  ");
	assert(signifix::format( 1e+00, signifix::PROPER) ==  "1.000  ");

	assert(signifix::format(-1e+00, signifix::SIGNED) == "-1.000  ");
	assert(signifix::format( 1e+00, signifix::SIGNED) == "+1.000  ");

	assert(signifix::format(-1e+00, signifix::SPACED) == "-1.000  ");
	assert(signifix::format( 1e+00, signifix::SPACED) == " 1.000  ");
}


tests/fp_category_safety.cxx
// This file is part of Signifix, see <https://qu1x.org/signifix>.
//
// Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
//
// Signifix is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License version 3
// as published by the Free Software Foundation on 29 June 2007.
//
// Signifix is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with Signifix. If not, see <https://www.gnu.org/licenses>.

#include <signifix/signifix.hxx>

#include <cassert>
#include <limits>
#include <cmath>

int main() {
	try {
		signifix::format(0.0);
		assert(false);
	}
	catch (signifix::lower_error& error) {
		assert(error.why() == 0.0);
	}
	catch (...) {
		assert(false);
	}

	try {
		signifix::format(-std::numeric_limits<double>::infinity());
		assert(false);
	}
	catch (signifix::upper_error& error) {
		assert(error.why() == -std::numeric_limits<double>::infinity());
	}
	catch (...) {
		assert(false);
	}

	try {
		signifix::format(std::numeric_limits<double>::infinity());
		assert(false);
	}
	catch (signifix::upper_error& error) {
		assert(error.why() == std::numeric_limits<double>::infinity());
	}
	catch (...) {
		assert(false);
	}

	try {
		signifix::format(std::numeric_limits<double>::quiet_NaN());
		assert(false);
	}
	catch (signifix::inval_error& error) {
		assert(std::isnan(error.why()));
	}
	catch (...) {
		assert(false);
	}
}


src/signifix/signifix.h
// This file is part of Signifix, see <https://qu1x.org/signifix>.
//
// Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
//
// Signifix is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License version 3
// as published by the Free Software Foundation on 29 June 2007.
//
// Signifix is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with Signifix. If not, see <https://www.gnu.org/licenses>.

//!
//! C header file of libsignifix.
//!
//! @file signifix.h
//!
#ifndef _SIGNIFIX_H
#define _SIGNIFIX_H

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

//!
//! The string buffer must be of at least this length.
//!
#define SIGNIFIX_BUFLEN 10

//!
//! Prefix only negative numbers, with '-'.
//!
#define SIGNIFIX_PROPER  0

//!
//! Additionally prefix positive numbers, with '+'.
//!

#define SIGNIFIX_SIGNED '+'

//!
//! Additionally prefix positive numbers, with ' '.
//!
#define SIGNIFIX_SPACED ' '

//!
//! Out of upper bound ±999.9 Y (≅ ±1e+27).
//!
#define SIGNIFIX_EUPPER -5

//!
//! Out of lower bound ±1.000 y (= ±1e-24).
//!
#define SIGNIFIX_ELOWER -4

//!
//! Invalid argument being not a number (NaN).
//!
#define SIGNIFIX_EINVAL -3

//!
//! Truncation due to undersized buffer.
//!
#define SIGNIFIX_ETRUNC -2

//!
//! Internal formatting error.
//!
#define SIGNIFIX_EPRINT -1

//!
//! Obtains the error string related to an error number.
//!
//! @param ERRNUM An error number returned by signifix_format().
//!
#define SIGNIFIX_ERRSTR(ERRNUM) signifix_errstr[(ERRNUM) + \
	sizeof signifix_errstr / sizeof signifix_errstr[0]]

//!
//! Error strings to be obtained by #SIGNIFIX_ERRSTR().
//!
const char* signifix_errstr[] = {
	"Out of upper bound ±999.9 Y (≅ ±1e+27)",
	"Out of lower bound ±1.000 y (= ±1e-24)",
	"Invalid argument being not a number (NaN)",
	"Truncation due to undersized buffer",
	"Internal formatting error",
};

//!
//! Number Formatter of Fixed Significance with Metric Unit Prefix.
//!
//! Formats a number to a string of a fixed number of significant figures and a
//! fixed number of resulting characters by automatically choosing the metric
//! unit prefix and appropriately adjusting the floating point precision.
//!
//! @param buffer  The buffer of the formatted string, allocated and freed by
//!                the caller. Must be of at least ::SIGNIFIX_BUFLEN bytes.
//! @param number  The number to be formatted.
//! @param plus    The plus sign character. Predefined values are
//!                ::SIGNIFIX_PROPER, ::SIGNIFIX_SIGNED and ::SIGNIFIX_SPACED.
//!
//! @return The string length in bytes of the formatted number if successful,
//!         otherwise a negative error number. The error string related to an
//!         error number can be obtained by #SIGNIFIX_ERRSTR(). Possible errors
//!         are #SIGNIFIX_EUPPER, #SIGNIFIX_ELOWER, #SIGNIFIX_EINVAL,
//!         #SIGNIFIX_ETRUNC, and #SIGNIFIX_EPRINT.
//!
int signifix_format(char* buffer, double number, char plus);

#endif


src/signifix/signifix.c
// This file is part of Signifix, see <https://qu1x.org/signifix>.
//
// Copyright (c) 2016 Rouven Spreckels <n3vu0r@qu1x.org>
//
// Signifix is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License version 3
// as published by the Free Software Foundation on 29 June 2007.
//
// Signifix is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with Signifix. If not, see <https://www.gnu.org/licenses>.

//!
//! C source file of libsignifix.
//!
//! @file signifix.c
//!
#ifndef _SIGNIFIX_
#define _SIGNIFIX_

#include "signifix.h"

#include <math.h>
#include <stdio.h>

int signifix_format(char* buffer, double number, char plus) {
	const double factor[] = {
		1e-24, 1e-21, 1e-18, 1e-15, 1e-12, 1e-09, 1e-06, 1e-03,
		1e-00,
		1e+03, 1e+06, 1e+09, 1e+12, 1e+15, 1e+18, 1e+21, 1e+24,
	};
	const char* symbol[] = {
		"y", "z", "a", "f", "p", "n", "µ", "m",
		" ",
		"k", "M", "G", "T", "P", "E", "Z", "Y",
	};
	char sign[2] = "-";
	if (number < 0.0)
		number = -number;
	else
		sign[0] = plus;
	int prefix;
	if (number < factor[8]) {
		if (number < factor[4]) {
			if (number < factor[2])
				prefix = number < factor[1] ? 0 : 1;
			else
				prefix = number < factor[3] ? 2 : 3;
		}
		else {
			if (number < factor[6])
				prefix = number < factor[5] ? 4 : 5;
			else
				prefix = number < factor[7] ? 6 : 7;
		}
	}
	else
	if (number < factor[9])
		prefix = 8;
	else {
		if (number < factor[13]) {
			if (number < factor[11])
				prefix = number < factor[10] ? 9 : 10;
			else
				prefix = number < factor[12] ? 11 : 12;
		} else {
			if (number < factor[15])
				prefix = number < factor[14] ? 13 : 14;
			else
				prefix = number < factor[16] ? 15 : 16;
		}
	}
	if (prefix != 8)
		number = number / factor[prefix];
	int strlen;
	const double middle = round(number * 1e+02);
	if (middle < 1e+04) {
		const double lower = round(number * 1e+03);
		if (lower < 1e+04) {
			if (lower < 1e+03)
				return SIGNIFIX_ELOWER;
			strlen = snprintf(buffer, SIGNIFIX_BUFLEN,
				"%s%.3f %s", sign, lower / 1e+03, symbol[prefix]);
		} else
			strlen = snprintf(buffer, SIGNIFIX_BUFLEN,
				"%s%.2f %s", sign, middle / 1e+02, symbol[prefix]);
	} else {
		const double upper = round(number * 1e+01);
		if (upper < 1e+04)
			strlen = snprintf(buffer, SIGNIFIX_BUFLEN,
				"%s%.1f %s", sign, upper / 1e+01, symbol[prefix]);
		else {
			if (++prefix == sizeof factor / sizeof factor[0])
				return isnan(number) ? SIGNIFIX_EINVAL : SIGNIFIX_EUPPER;
			strlen = snprintf(buffer, SIGNIFIX_BUFLEN,
				"%s1.000 %s", sign, symbol[prefix]);
		}
	}
	if (strlen < 0)
		return SIGNIFIX_EPRINT;
	if (strlen >= SIGNIFIX_BUFLEN)
		return SIGNIFIX_ETRUNC;
	return strlen;
}

#endif
